engine/src/contract_methods/connector.rs

```diff
@@ -0,0 +1,629 @@
+        Ok(address)
+    })
+}
+
+#[named]
+pub fn exit_to_near_precompile_callback<I: IO + Copy, E: Env, H: PromiseHandler>(
+    io: I,
+    env: &E,
+    handler: &mut H,
+) -> Result<Option<SubmitResult>, ContractError> {
+    with_hashchain(io, env, function_name!(), |io| {
+        let state = state::get_state(&io)?;
+        require_running(&state)?;
+        env.assert_private_call()?;
+
+        // This function should only be called as the callback of
+        // exactly one promise.
+        if handler.promise_results_count() != 1 {
+            return Err(errors::ERR_PROMISE_COUNT.into());
+        }
+
+        let args: ExitToNearPrecompileCallbackArgs = io.read_input_borsh()?;
+
+        let maybe_result = if let Some(PromiseResult::Successful(_)) = handler.promise_result(0) {
+            if let Some(args) = args.transfer_near {
+                let action = PromiseAction::Transfer {
+                    amount: Yocto::new(args.amount),
+                };
+                let promise = PromiseBatchAction {
+                    target_account_id: args.target_account_id,
+                    actions: vec![action],
+                };
+
+                // Safety: this call is safe because it comes from the exit to near precompile, not users.
+                // The call is to transfer the unwrapped wNEAR tokens.
+                let promise_id = unsafe { handler.promise_create_batch(&promise) };
+                handler.promise_return(promise_id);
+            }
+
+            None
+        } else if let Some(args) = args.refund {
+            // Exit call failed; need to refund tokens
+            let refund_result = engine::refund_on_error(io, env, state, &args, handler)?;
+
+            if !refund_result.status.is_ok() {
+                return Err(errors::ERR_REFUND_FAILURE.into());
+            }
+
+            Some(refund_result)
+        } else {
+            None
+        };
+
+        Ok(maybe_result)
+    })
+}
+
+pub fn ft_transfer<I: IO + Env + Copy + PromiseHandler, E: Env>(
+    mut io: I,
+    env: &E,
+) -> Result<(), ContractError> {
+    require_running(&state::get_state(&io)?)?;
+    env.assert_one_yocto()?;
+    let input = read_json_args(&io).and_then(|args: FtTransferArgs| {
+        serde_json::to_vec(&(
+            env.predecessor_account_id(),
+            args.receiver_id,
+            args.amount,
+            args.memo,
+        ))
+        .map_err(Into::<ParseArgsError>::into)
+    })?;
+
+    let promise_arg = PromiseCreateArgs {
+        target_account_id: get_connector_account_id(&io)?,
+        method: "engine_ft_transfer".to_string(),
+        args: input,
+        attached_balance: ONE_YOCTO,
+        attached_gas: calculate_attached_gas(env),
+    };
+    let promise_id = unsafe { io.promise_create_call(&promise_arg) };
+    io.promise_return(promise_id);
+
+    Ok(())
+}
+
+pub fn ft_transfer_call<I: IO + Env + Copy, E: Env, H: PromiseHandler>(
```

Comment:
I wonder if there is a way to generalize the NEP-141 functions that are now simply forwarding to the external connector. They all have the same structure of reading the args and constructing a promise. This is not a blocking comment, just a suggestion that might reduce some minor repetition in the code.

Reply:
Yeah. It's a good point.

Reply:
Done in ec547c8